transport/api/src/lib.rs

```diff
@@ -299,26 +208,146 @@ where
+            constants::MAXIMUM_NETWORK_UPDATE_EVENT_QUEUE_SIZE,
+        );
+
+        // manual ping
+        let (ping_tx, ping_rx) = futures::channel::mpsc::unbounded::<(PeerId, ControlMessage)>();
+        let (pong_tx, pong_rx) =
+            futures::channel::mpsc::unbounded::<(PeerId, std::result::Result<(ControlMessage, String), ()>)>();
+
+        let ping_cfg = PingConfig {
+            timeout: proto_cfg.heartbeat.timeout,
+            ..PingConfig::default()
+        };
+
+        let ping: Ping<adaptors::ping::PingExternalInteractions<T>> = Ping::new(
+            ping_cfg,
+            ping_tx,
+            pong_rx,
+            adaptors::ping::PingExternalInteractions::new(
+                network.clone(),
+                db.clone(),
+                channel_graph.clone(),
+                network_events_tx.clone(),
+            ),
+        );
+
+        // heartbeat
+        let (hb_ping_tx, hb_ping_rx) = futures::channel::mpsc::unbounded::<(PeerId, ControlMessage)>();
+        let (hb_pong_tx, hb_pong_rx) = futures::channel::mpsc::unbounded::<(
+            libp2p::identity::PeerId,
+            std::result::Result<(ControlMessage, String), ()>,
+        )>();
+
+        let ping_cfg = PingConfig {
+            timeout: proto_cfg.heartbeat.timeout,
+            ..PingConfig::default()
+        };
+
+        let hb_pinger = Ping::new(
+            ping_cfg,
+            hb_ping_tx,
+            hb_pong_rx,
+            adaptors::ping::PingExternalInteractions::new(
+                network.clone(),
+                db.clone(),
+                channel_graph.clone(),
+                network_events_tx,
+            ),
+        );
+        let heartbeat = Heartbeat::new(
+            hb_cfg,
+            hb_pinger,
+            core_network::heartbeat::HeartbeatExternalInteractions::new(network.clone()),
+        );
+
+        let pkt_sender = packet_actions.writer();
+        let ticket_aggregate_actions = ticket_aggregation.writer();
+
+        let identity: libp2p::identity::Keypair = (me).into();
+
+        let swarm_loop = SwarmEventLoop::new(
+            identity.clone(),
+            my_multiaddresses.clone(),
+            network_events_rx,
+            indexer_update_rx,
+            ack_actions,
+            packet_actions,
+            ticket_aggregation,
+            core_p2p::api::HeartbeatRequester::new(hb_ping_rx),
+            core_p2p::api::HeartbeatResponder::new(hb_pong_tx),
+            core_p2p::api::ManualPingRequester::new(ping_rx),
+            core_p2p::api::HeartbeatResponder::new(pong_tx),
+        );
+
+        // : Some((heartbeat, swarm_loop))
+
+        let (push_tx, push_rx) = async_channel::bounded::<HoprSwarmArgs<T>>(1);
+        let (pull_tx, pull_rx) =
+            async_channel::bounded::<std::collections::HashMap<HoprTransportProcess, JoinHandle<()>>>(1);
+
+        spawn(async move {
```

Comment:
I assume we don't need to await this JoinHandle, since the receiver inside gets closed, right?

Reply:
Correct, once all TX are closed, the RX will yield None at which point the internal loop will terminate.